### PR TITLE
Highcharts support for theming.

### PR DESCRIPTION
**Pull request by yanlipnican**
Added support for themes in highcharts trough ITheme properties.

JIRA: BB-2815

Out of scope:
 - heatmap

---

Supported PR commands:

| Command                | Description            |
| ---------------------- | ---------------------- |
| `ok to test`           | Re-run standard checks |
| `extended test`        | BackstopJS tests       |
| `extended check sonar` | SonarQube tests        |

---

# PR Checklist

-   [ ] commit messages adhere to the [commit message guidelines](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#what-should-the-commits-look-like)
-   [ ] review was done by a Code owner [if necessary](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-tell-if-my-pull-request-needs-approval-by-a-code-owner) (if you think it is not necessary, explain the reasoning in the description or in a comment)
-   [ ] `check` passes
-   [ ] `check-extended` passes
-   [ ] `rush change` [was run if applicable](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-describe-my-changes-for-the-changelog)

### CODE REVIEW

**Comment by gdgate** [comment]
Merge Failed.

Zuul merger could not merge this change into the base branch. This is most likely caused by merge conflicts. Please rebase the change and upload the rebased version. In case of further errors, contact the zuul administrator.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/2576/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/4206/ : FAILURE in 15m 39s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/2603/ : SUCCESS in 9s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/4234/ : FAILURE in 12m 14s

**Comment by gdgate** [comment]
Merge Failed.

Zuul merger could not merge this change into the base branch. This is most likely caused by merge conflicts. Please rebase the change and upload the rebased version. In case of further errors, contact the zuul administrator.

File: libs/sdk-ui-vis-commons/styles/scss/legend.scss [hunk 1]
```diff
@@ -98,28 +96,31 @@ $responsive-legend: "only screen and (max-width: 767px)";
...
             @include active-states-and-self {
                 text-decoration: none;
-                background-color: $legend-paging-background;
+                background-color: var(--gd-palette-complementary-2, $legend-paging-background);
 
                 // refactor?
                 // stylelint-disable-next-line max-nesting-depth
                 &::before {
-                    color: $gd-color-dark;
+                    color: var(--gd-palette-complementary-7, $default-gd-color-dark);
```
Comment (yanlipnican) [inline libs/sdk-ui-vis-commons/styles/scss/legend.scss:104]: There is different complementary shade in `$gd-color-dark`

File: libs/sdk-ui-charts/src/highcharts/chartTypes/_chartCreators/commonConfiguration.ts [hunk 2]
```diff
@@ -55,6 +60,7 @@ function getThemedConfiguration(theme: ITheme): any {
             enabled: false,
         },
         drilldown: {
+            // TODO drilling
```
Comment (scavnickyj) [inline libs/sdk-ui-charts/src/highcharts/chartTypes/_chartCreators/commonConfiguration.ts:63]: TODO

File: libs/sdk-ui-charts/src/highcharts/chartTypes/scatterPlot/scatterConfiguration.ts [hunk 3]
```diff
@@ -1,63 +1,65 @@
...
+        chart: {
+            type: "scatter",
+        },
+        plotOptions: {
+            scatter: {
+                // dataLabels: {
+                //     style: {
+                //         color: theme?.palette?.complementary?.shade9 ?? styleVariables.gdColorText,
+                //     },
+                // },
```
Comment (scavnickyj) [inline libs/sdk-ui-charts/src/highcharts/chartTypes/scatterPlot/scatterConfiguration.ts:24]: What's this for?

File: libs/sdk-ui-charts/src/highcharts/constants/label.ts [hunk 4]
```diff
@@ -1,14 +1,14 @@
...
 export const WHITE_LABEL: Highcharts.CSSObject = {
-    color: "#ffffff",
-    textShadow: "0 0 1px #000000",
+    color: "var(--gd-palette-complementary-9, #ffffff)",
+    textShadow: "0 0 1px var(--gd-shadow-color, #000000)",
 };
 
 export const BLACK_LABEL: Highcharts.CSSObject = {
-    color: "#000000",
+    color: "var(--gd-palette-complementary-9, #000000)",
```
Comment (scavnickyj) [inline libs/sdk-ui-charts/src/highcharts/constants/label.ts:11]: Are these complementary indices correct? It's strange that both black and white map to 9...

File: libs/sdk-ui-charts/styles/scss/charts.scss [hunk 5]
```diff
@@ -52,6 +52,7 @@
     }
     // stylelint-disable declaration-no-important
     .gd-axis-label-drilling-disabled & {
+        // TODO: Drilling
```
Comment (scavnickyj) [inline libs/sdk-ui-charts/styles/scss/charts.scss:55]: TODO

File: libs/sdk-ui-charts/styles/scss/charts.scss [hunk 6]
```diff
@@ -88,5 +89,5 @@
 
 .gd-contrast-label,
 .gd-contrast-label text {
-    fill: var(--gd-chart-backgroundColor-contrast, "#000");
+    fill: var(--gd-palette-complementary-9, "#000");
```
Comment (scavnickyj) [inline libs/sdk-ui-charts/styles/scss/charts.scss:92]: Shouldn't this be used as a fallback to `--gd-chart-backgroundColor-contrast` ?

File: libs/sdk-ui-charts/styles/scss/headline.scss [hunk 7]
```diff
@@ -1,6 +1,7 @@
 // (C) 2007-2018 GoodData Corporation
-@import "~@gooddata/goodstrap/lib/core/styles/typo-mixins";
-@import "~@gooddata/goodstrap/lib/core/styles/layout";
+@import "~@gooddata/sdk-ui-kit/styles/scss/variables";
```
Comment (scavnickyj) [inline libs/sdk-ui-charts/styles/scss/headline.scss:2]: Not 100% sure about this, but maybe we can loose the variables import, as it is imported inside typo-mixins below. I'm no sass expert so I don't know if it makes any difference..

File: libs/sdk-ui-charts/styles/scss/headline.scss [hunk 8]
```diff
@@ -31,14 +32,15 @@
             }
         }
         .headline-value--empty {
-            color: var(--gd-kpi-primaryMeasureColor, $gd-color-disabled);
+            color: var(--gd-kpi-primaryMeasureColor, $default-gd-color-disabled);
```
Comment (scavnickyj) [inline libs/sdk-ui-charts/styles/scss/headline.scss:35]: Is this intentional? Do we need fixed default here?

File: libs/sdk-ui-charts/styles/scss/headline.scss [hunk 9]
```diff
@@ -65,7 +67,7 @@
     }
 
     .headline-value--empty {
-        color: var(--gd-kpi-secondaryInfoColor, $gd-color-disabled);
+        color: var(--gd-kpi-secondaryInfoColor, $default-gd-color-disabled);
```
Comment (scavnickyj) [inline libs/sdk-ui-charts/styles/scss/headline.scss:70]: Same as above..

File: libs/sdk-ui-charts/styles/scss/tooltip.scss [hunk 10]
```diff
@@ -1,28 +1,17 @@
 // (C) 2007-2020 GoodData Corporation
-@import "~@gooddata/goodstrap/lib/Bubble/styles/mixins";
-@import "~@gooddata/goodstrap/lib/core/styles/variables";
+@import "~@gooddata/sdk-ui-kit/styles/scss/variables";
+@import "~@gooddata/sdk-ui-kit/styles/scss/mixins";
 @import "variables";
```
Comment (scavnickyj) [inline libs/sdk-ui-charts/styles/scss/tooltip.scss:4]: I believe that sdk-ui-kit variables are being imported from `@import "variables";` line.

File: libs/sdk-ui-charts/styles/scss/tooltip.scss [hunk 11]
```diff
@@ -1,28 +1,17 @@
...
-    .tt-tooltip {
-        margin-top: 2px;
-        margin-bottom: 14px;
-        font-size: 14px;
-    }
+    color: var(--gd-palette-complementary-6, $default-gd-color-dark);
+    background-color: var(
+        --gd-chart-tooltipBackgroundColor,
+        var(--gd-palette-complementary-0, transparentize($default-gd-color-white, 0.05))
+    );
```
Comment (scavnickyj) [inline libs/sdk-ui-charts/styles/scss/tooltip.scss:14]: I'm missing previous variables as fallbacks here. Is it okay? Also do we need explicitly need `$default-gd-color-dark`?

File: libs/sdk-ui-charts/styles/scss/tooltip.scss [hunk 12]
```diff
@@ -41,8 +30,12 @@
     padding: 10px 9px 9px;
     border-radius: 3px;
     border: 1px solid;
-    border-color: var(--gd-chart-textColor-light, transparentize($bubble-light-border-base, 0.3));
-    box-shadow: 0 1px 3px 0 rgba(20, 56, 93, 0.1), 0 2px 9px 0 rgba(20, 56, 93, 0.15);
+    border-color: var(
+        --gd-chart-tooltipBorderColor,
```
Comment (scavnickyj) [inline libs/sdk-ui-charts/styles/scss/tooltip.scss:34]: I'm missing previous variable as fallback here. Is it ok?

File: libs/sdk-ui-theme-provider/src/cssProperties.ts [hunk 13]
```diff
@@ -294,11 +294,12 @@ const getFormDerivedColors = (palette: IThemePalette): CssProperty[] => [
...
-    getCssProperty(
-        "chart-backgroundColor-base-t05",
-        chart?.backgroundColor?.base && transparentize(0.05, chart?.backgroundColor?.base),
-    ),
+// TODO: Not needed
+const getChartDerivedColors = (_chart: IThemeChart): CssProperty[] => [
+    // getCssProperty(
+    //     "chart-backgroundColor-t05",
+    //     chart?.backgroundColor && transparentize(0.05, chart?.backgroundColor),
+    // ),
```
Comment (scavnickyj) [inline libs/sdk-ui-theme-provider/src/cssProperties.ts:302]: TODO, remove?

File: libs/sdk-ui-vis-commons/styles/scss/legend.scss [hunk 14]
```diff
@@ -39,7 +38,6 @@ $responsive-legend: "only screen and (max-width: 767px)";
 
 .viz-legend {
     display: flex;
-    background-color: var(--gd-chart-backgroundColor-base, $gd-color-white);
```
Comment (scavnickyj) [inline libs/sdk-ui-vis-commons/styles/scss/legend.scss:42]: Wouldn't this change current legend state?

File: libs/sdk-ui-charts/styles/scss/tooltip.scss [hunk 15]
```diff
@@ -86,6 +79,7 @@
     margin-right: -1em;
     padding-right: 1em;
     word-wrap: break-word;
+    color: var(--gd-chart-tooltipValueColor, $gd-color-dark);
```
Comment (scavnickyj) [inline libs/sdk-ui-charts/styles/scss/tooltip.scss:82]: Wouldn't this change current tooltip state?

**Comment by gdgate** [comment]
Merge Failed.

Zuul merger could not merge this change into the base branch. This is most likely caused by merge conflicts. Please rebase the change and upload the rebased version. In case of further errors, contact the zuul administrator.

**Comment by yanlipnican** [comment]
extended test

**Comment by yenkins** [comment]
Link to Storybook test result: https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-backstop-tests/703/artifact/libs/sdk-ui-tests/backstop/output/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/2643/ : SUCCESS in 19s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-backstop-tests/703/ : SUCCESS in 23m 13s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/2642/ : SUCCESS in 10s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/4274/ : SUCCESS in 15m 37s

**Review by scavnickyj — APPROVED** [review]

**Comment by yenkins** [comment]
### Sonar scan result
 - No new bug found. Congratulation!

More detail, see in https://sonarqube-dev.intgdc.com/dashboard?id=gooddata-ui-sdk-gate-PR1069

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-fossa-scanning-zuul-el7/792/ : FAILURE in 3m 18s (non-voting)
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-merge-compliance-check/1056/ : SUCCESS in 19s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/4277/ : SUCCESS in 25m 03s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-sonarqube-scan-zuul-el7/157/ : SUCCESS in 9m 41s (non-voting)

**Pull request merged**